src/masternodes/mn_checks.cpp

```diff
@@ -1029,6 +1114,443 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
         return mnview.SetOracleData(obj.oracleId, obj.timestamp, obj.tokenPrices);
     }
 
+    Res operator()(const CICXCreateOrderMessage& obj) const {
+        auto res = CheckICXTx();
+        if (!res)
+            return res;
+
+        CICXOrderImplemetation order;
+        static_cast<CICXOrder&>(order) = obj;
+
+        order.creationTx = tx.GetHash();
+        order.creationHeight = height;
+
+        if (order.idToken.v==std::numeric_limits<uint32_t>::max())
+            return Res::Err("%s: %s", __func__, "token assets must be specified");
+        if (order.chain.empty())
+            return Res::Err("%s: %s", __func__, "chain assets must be specified");    
+        if (order.orderType != CICXOrder::TYPE_INTERNAL && order.orderType != CICXOrder::TYPE_EXTERNAL)
+            return Res::Err("invalid order type!");
+
+        if (order.orderType == CICXOrder::TYPE_INTERNAL)
+        {
+            auto tokenFrom = mnview.GetToken(order.idToken);
+            if (!tokenFrom)
+                return Res::Err("%s: %s", __func__, "tokenFrom (" +  tokenFrom->CreateSymbolKey(order.idToken) + ") does not exist!");
+
+            if (order.ownerAddress.empty())
+                return Res::Err("ownerAddress must be not-null!");
+
+            // subtract the balance from tokenFrom to dedicate them for the order
+            CScript txidAddr(order.creationTx.begin(), order.creationTx.end());
+            res = ModifyTokenBalance(order.idToken, order.amountFrom, order.ownerAddress, txidAddr);
+            if (!res.ok)
+                return Res::Err("%s: %s", __func__, res.msg);
+        }
+        else if (order.orderType == CICXOrder::TYPE_EXTERNAL)
+        {
+            auto tokenTo = mnview.GetToken(order.idToken);
+            if (!tokenTo)
+                return Res::Err("%s: %s", __func__, "tokenTo id (" +  tokenTo->CreateSymbolKey(order.idToken) + ") does not exist!");
+        }
+
+        if (order.amountFrom == 0)
+            return Res::Err("order amountFrom must be greater than 0!");
+        if (order.amountToFill != order.amountFrom)
+            return Res::Err("order amountToFill does not equal to amountFrom!");
+        if (order.orderPrice == 0)
+            return Res::Err("order price must be greater than 0!");
+        if (order.expiry == 0)
+            return Res::Err("order expiry must be greater than 0!");
+
+        return mnview.ICXCreateOrder(order);
+    }
+
+    Res operator()(const CICXMakeOfferMessage& obj) const {
```

Comment:
Since CloseOffer requires auth of create offer TX vout 1, we should probably make sure this TX has the correct auth input with HasAuth set to receive or owner address.

Reply:
Already in mn_checks.cpp